### ISSUE TITLE
Harden security

### ISSUE THREAD

**Issue by aminb** [hlissner/.emacs.d#72]
See [this](https://glyph.twistedmatrix.com/2015/11/editor-malware.html) blog post.

I've adapted the recommendations for myself (I use Arch) [here](https://github.com/aminb/.emacs.d/commit/a7147cdfe0c8ab21c41b28095efbd3c382c6381c). That repo isn't Doom-based, but there's nothing specific to my Emacs setup in that commit.

**Comment by hlissner** [comment]
Excellent advice. I'll look into incorporating these changes asap. Thank you for sharing it!

**Comment by aminb** [comment]
Cheers!

**Comment by hlissner** [comment]
Hmm, looks like I'll have to compromise if I want `org-contrib-plus` from the OrgMode ELPA repo (which doesn't offer secure access). 

If I put `http://orgmode.org/elpa/` into `package-archives` I suffer the possiblity of a compromised local archive for many packages (a lot of tainted package sources; many vectors of attack).

I suppose I could mitigate the damage with quelpa, by fetching that one package directly from their git repo (via HTTP, but presents only one vector of attack).

**Comment by aminb** [comment]
Yeah that's a good point, and I've noticed it too.

I reached out to the maintainer of Org mode today about making orgmode.org available via https. Let's see what happens.

In the mean time, I've noticed that [req-package](https://github.com/edvorg/req-package) allows specifying a [provider](https://github.com/edvorg/req-package/blob/develop/req-package-providers.el) using `:loader`. I might try to use that or a similar strategy to explicitly specify which package archive a package comes from, and only fetch `org-plus-contrib` from the Org ELPA.

P.S. was not setting `tls-program` in e678d13 intentional?

**Comment by hlissner** [comment]
> I reached out to the maintainer of Org mode today about making orgmode.org available via https.

Here's to hoping!

>P.S. was not setting tls-program in e678d13 intentional?

Yes, but only temporarily while I looked into whether I needed openssl in that list or not (I'm leaning towards not). At the very least, the second gnutls-cli command should go, since we shouldn't trust the ssl3 protocol.

**Comment by aminb** [comment]
Good catch on SSLv3!

**Comment by hlissner** [comment]
Alright, I added `tls-program`. I kept openssl for backwards compatibility. Outright failure to verify will prevent package installation/update from those sources (the one exception being org-plus-contrib).

I added `bin/doctor` to help diagnose issues with the environment. It will report if gnutls-cli/openssl is missing, or if openssl is too old (i.e. doesn't support TLS 1.1+). But these are just warnings. I ultimately leave it up to the user.

With that, I think I can consider this resolved.
